File: js/Samples/Tetris_GoodAbstractionLevel.js
class Field {
    constructor(width, height, filledCellsLineByLine, score = 0) {
        this.width = width;
        this.height = height;
        this.score = score;
        this.filledCellsLineByLine = filledCellsLineByLine;
    }

    clearFullLines() {
        const notFullLines = this.getAllNotFullLines();
        const clearedLinesCount = this.height - notFullLines.length;
        const newLinesArray = this.createNewLinesArray(clearedLinesCount, notFullLines);
        return new Field(this.width, this.height, newLinesArray, this.score + clearedLinesCount);
    }

    createNewLinesArray(emptyLinesCount, nonEmptyLines) {
        const emptyLines = Array.from(Array(emptyLinesCount));
        return emptyLines.concat(nonEmptyLines);
    }

    getAllNotFullLines()
    {
        return this.filledCellsLineByLine.filter(line =>
            line.filter(item => Boolean(item)).length !== this.width);
    }
}

File: python/Samples/tetris_good_abstraction_level.py
class Field:
    def __init__(self, width, height, filled_cells_line_by_line, score=0):
        self.width = width
        self.height = height
        self.score = score
        self.filled_cells_line_by_line = filled_cells_line_by_line

    def clear_full_lines(self):
        not_full_lines = self.get_all_not_full_lines()
        cleared_lines_count = self.height - len(not_full_lines)
        new_lines_array = self.create_new_lines_array(
            cleared_lines_count, not_full_lines
        )
        print(new_lines_array)
        return Field(
            self.width, self.height, new_lines_array, self.score + cleared_lines_count
        )

    def create_new_lines_array(self, empty_lines_count, non_empty_lines):
        empty_lines = [[False, ] * self.width] * empty_lines_count
        return non_empty_lines + empty_lines

    def get_all_not_full_lines(self):
        return [
            line for line in self.filled_cells_line_by_line if sum(line) != self.width
        ]
